### PR TITLE
controllers: avoid double copy of the rest config

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
adding a user agent changes the original config, which might be used by other clients.
## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/server/controllers.go [hunk 1]
```diff
@@ -954,13 +954,11 @@ func (s *Server) installSyncTargetController(ctx context.Context, config *rest.C
...
 
 func (s *Server) installKubeQuotaController(
-	ctx context.Context,
 	config *rest.Config,
 	server *genericapiserver.GenericAPIServer,
 ) error {
 	controllerName := "kcp-kube-quota-controller"
-	config = rest.AddUserAgent(config, controllerName)
-
+	config = kcpclienthelper.NewClusterConfig(rest.AddUserAgent(rest.CopyConfig(config), controllerName))
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/server/controllers.go:961]: it was cluster aware befoe, it didn't make a copy of the rest config :)

[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/server/controllers.go:961]: actually sorry, it looks like it didn't copy the rest config and it wasn't cluster aware.

**Comment by p0lyn0mial** [comment]
also, I've just realized that `kcpclienthelper.NewClusterConfig` already does a copy, so i changed the way the controllers add a user agent to avoid double copy, ptal

File: pkg/server/controllers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/server/controllers.go]: @ncdc is it okay to run the quota controller with the cluster-aware RT?

**Comment by stevekuznetsov** [comment]
While this achieves what you want, I think it might be useful here to be *slightly* more explicit - as it stands, this depends implicitly on a private implementation detail of a helper in another repo for correctness, and someone who has not seen this PR will not know why there's only one correct way to do this. Perhaps changing the helper not to copy would be correct, AFAICT other `*rest.Config` mutators don't copy, do they?

**Comment by p0lyn0mial** [comment]
> as it stands, this depends implicitly on a private implementation detail of a helper in another repo for the correctness

can we introduce a breaking change to https://github.com/kcp-dev/apimachinery repo? I'm not sure who uses this repo, if it is only kcp then let's change it.

**Comment by stevekuznetsov** [comment]
Downstream users have to import and use it as well. That's a shame. @ncdc @sttts any thoughts? At this point we're breaking so frequently it's not IMHO that big of a deal, especially since the normal user expectation is that helpers are not copying your config for you anyway.

File: pkg/server/controllers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/server/controllers.go]: I'll need some time to review and see if it has any negative effects (or, if the e2es pass, presumably it doesn't 😄).

Maybe don't switch to NewClusterConfig for quota in this PR and we can do it separately?

**Comment by ncdc** [comment]
We're probably at a point where we need to tag kcp-dev/apimachinery and make breaking changes in new versions.

I kinda feel like there's something substantively different between adding a user agent and changing cfg.Host... right?

Maybe we have NewClusterConfig not make a copy, since all it does is add a RoundTripper that is only operative if the context has a cluster set. But keep it so ConfigWithCluster makes a copy internally, because of the modification to cfg.Host?

**Comment by ncdc** [comment]
>  as it stands, this depends implicitly on a private implementation detail of a helper in another repo for correctness

I'd rather that we make a copy at the callsite where we know we don't want to mutate the original config, i.e. in the code in this repo.

**Comment by stevekuznetsov** [comment]
Good point re: overwriting the host entirely on the WithCluster call. So perhaps a new tag on the machinery repo (maybe we want to co-locate that with the changes to k8s fork that would allow those calls to live in `logicalcluster` for simplicity?). We can merge this PR for now if we've got the future sketched out

**Comment by ncdc** [comment]
>We can merge this PR for now

I actually wonder if we shouldn't, since making this change removes the copy where we need it. Would it make more sense to not merge, and instead modify apimachinery so NewClusterConfig doesn't make a copy?

**Comment by stevekuznetsov** [comment]
My thought was that we would use the implict copy until we bumped the dep, but yes if you'd prefer to make the change in apimachinery first, I'm all for it. I didn't want @p0lyn0mial to feel blocked though - @p0lyn0mial do you mind doing the work in that repo (and/or the k8s fork work) before we finish up this PR, or do you want me to handle some of that?

**Comment by p0lyn0mial** [comment]
I've decided to change both `NewClusterConfig` and `ConfigWithCluster` functions, PTAL  - https://github.com/kcp-dev/apimachinery/pull/52

**Comment by p0lyn0mial** [comment]
alright, I've updated the apimachinery repo and replaced references to old methods, ptal

File: pkg/server/server.go [hunk 2]
```diff
@@ -446,7 +446,7 @@ func (s *Server) Run(ctx context.Context) error {
 	}
 
 	if s.Options.Controllers.EnableAll || enabled.Has("quota") {
-		if err := s.installKubeQuotaController(ctx, controllerConfig, delegationChainHead); err != nil {
+		if err := s.installKubeQuotaController(controllerConfig, delegationChainHead); err != nil {
```
Comment (stevekuznetsov) [inline pkg/server/server.go:449]: I'm going to have to add this right back in this week for getting the logger from the context, do you mind not removing it for now?

Comment (p0lyn0mial) [reply pkg/server/server.go:449]: sure, no problem.

Comment (p0lyn0mial) [reply pkg/server/server.go]: done.

**Pull request merged**